### PR TITLE
Do not count MoE tokens during evaluation

### PR DESCRIPTION
**Pull request by taking-lying-flat**
`MoE.forward()` currently updates `tokens_per_expert_E` in eval mode. Validation runs after the optimizer hook has consumed and cleared the training counts, so validation routing counts can leak into the next step's `expert_bias_E` update. `torch.no_grad()` does not prevent this buffer mutation.

Only accumulate the counter while the module is in training mode. Activation-checkpoint recomputation still runs in training mode, so its existing behavior is unchanged.

Added a CPU regression test that first records training counts, then verifies an eval forward leaves them unchanged.

Tests:

- `python -m pytest -q tests/unit_tests/test_moe.py tests/unit_tests/test_optimizer_param_groups.py tests/unit_tests/test_inference_moe.py tests/unit_tests/test_gpt_oss_moe.py` (28 passed)
- File-scoped pre-commit hooks for the changed files
- Pyrefly on `torchtitan/models/common/moe.py` (0 errors)

### CODE REVIEW

**Review by tianyu-l — APPROVED** [review]
Sounds legit. Let's stop bleeding and worry about inference metrics later. @felipemello1

**Review by acisseJZhong — APPROVED** [review]

**Comment by acisseJZhong** [comment]
lgtm! we are not logging tokens_per_expert_E during training and inference at all right?
